ClientLib/src/main/java/com/o3dr/services/android/lib/drone/companion/solo/tlv/TLVMessageParser.java

```diff
@@ -132,36 +138,44 @@
                         final short gimbalRate = packetBuffer.getShort();
                         final short battery = packetBuffer.getShort();
 
-                        packetList.add(new ControllerMessageInputReport(timestamp, gimbalY, gimbalRate, battery));
+                        packet = new ControllerMessageInputReport(timestamp, gimbalY, gimbalRate, battery);
                         break;
                     }
 
                     case TYPE_SOLO_GOPRO_SET_REQUEST: {
                         @SoloGoproConstants.RequestCommand final short command = packetBuffer.getShort();
                         @SoloGoproConstants.RequestCommandValue final short value = packetBuffer.getShort();
-                        packetList.add(new SoloGoproSetRequest(command, value));
+                        packet = new SoloGoproSetRequest(command, value);
                         break;
                     }
 
                     case TYPE_SOLO_GOPRO_RECORD: {
                         @SoloGoproConstants.RecordCommand final int command = packetBuffer.getInt();
-                        packetList.add(new SoloGoproRecord(command));
+                        packet = new SoloGoproRecord(command);
                         break;
                     }
 
                     case TYPE_SOLO_GOPRO_STATE: {
-                        packetList.add(new SoloGoproState(packetBuffer));
+                        packet = new SoloGoproState(packetBuffer);
                         break;
                     }
 
                     case TYPE_SOLO_GOPRO_REQUEST_STATE: {
-                        packetList.add(new SoloGoproRequestState());
+                        packet = new SoloGoproRequestState();
                         break;
                     }
 
                     default:
                         break;
                 }
+
+                if (packet != null && packet.getMessageLength() == messageLength) {
+                    packetList.add(packet);
+                } else {
+                    packetBuffer.reset();
+                    packetBuffer.get(new byte[messageLength]);
```

Comment:
@chaviw use [`ByteBuffer#position(int)`](https://developer.android.com/reference/java/nio/Buffer.html#position(int)) to update the position of the `packetBuffer`.
You should always be wary of object allocation in a loop..